File: Apps/Flutter/lib/src/localization.dart
import 'models.dart';

Map<String, String> parseTomlStrings(String text) {
  final values = <String, String>{};
  final lines = text.split(RegExp(r'\r?\n'));
  var index = 0;
  while (index < lines.length) {
    final rawLine = lines[index];
    final line = rawLine.trim();
    index += 1;
    if (line.isEmpty || line.startsWith('#')) {
      continue;
    }
    final equals = _findUnescapedEquals(line);
    if (equals < 0) {
      throw FormatException('Invalid localization TOML: $rawLine');
    }
    final key = _unquoteKey(line.substring(0, equals).trim());
    var rawValue = line.substring(equals + 1).trimLeft();
    if (rawValue.startsWith('"""')) {
      rawValue = rawValue.substring(3);
      final collected = <String>[];
      final sameLineEnd = rawValue.indexOf('"""');
      if (sameLineEnd >= 0) {
        collected.add(rawValue.substring(0, sameLineEnd));
      } else {
        collected.add(rawValue);
        var foundEnd = false;
        while (index < lines.length) {
          final nextLine = lines[index];
          index += 1;
          final end = nextLine.indexOf('"""');
          if (end >= 0) {
            collected.add(nextLine.substring(0, end));
            foundEnd = true;
            break;
          }
          collected.add(nextLine);
        }
        if (!foundEnd) {
          throw FormatException(
              'Unterminated multiline localization string: $key');
        }
      }
      if (collected.isNotEmpty && collected.first.isEmpty) {
        collected.removeAt(0);
      }
      if (collected.isNotEmpty && collected.last.isEmpty) {
        collected.removeLast();
      }
      values[key] = collected.join('\n');
      continue;
    }
    if (!rawValue.startsWith('"')) {
      throw FormatException('Invalid localization TOML: $rawLine');
    }
    final closing = _findClosingQuote(rawValue);
    if (closing < 0) {
      throw FormatException('Invalid localization TOML: $rawLine');
    }
    values[key] = _unescapeTomlString(rawValue.substring(1, closing));
  }
  return values;
}

BundleManifest localizeManifest(
  BundleManifest manifest,
  Map<String, String> table,
) =>
    manifest.copyWith(
      displayName: localized(manifest.displayName, table),
      summary: localized(manifest.summary, table),
      setup: SetupSpec(
        steps: manifest.setup.steps
            .map((step) => SetupStepSpec(
                  id: step.id,
                  label: localized(step.label, table),
                  kind: step.kind,
                  value: step.value,
                  workingDirectory: step.workingDirectory,
                  arguments: step.arguments,
                  environment: step.environment,
                  optional: step.optional,
                ))
            .toList(),
      ),
      pages: manifest.pages.map((page) => _localizePage(page, table)).toList(),
    );

String localized(String value, Map<String, String> table) =>
    table[value] ?? value;

String? localizedOptional(String? value, Map<String, String> table) =>
    value == null ? null : localized(value, table);

BundlePage _localizePage(BundlePage page, Map<String, String> table) =>
    BundlePage(
      id: page.id,
      title: localized(page.title, table),
      summary: localized(page.summary, table),
      iconName: page.iconName,
      iconEmoji: page.iconEmoji,
      sidebarGroup: localizedOptional(page.sidebarGroup, table),
      sections: page.sections
          .map((section) => _localizeSection(section, table))
          .toList(),
    );

PageSection _localizeSection(PageSection section, Map<String, String> table) =>
    PageSection(
      id: section.id,
      title: localizedOptional(section.title, table),
      summary: localizedOptional(section.summary, table),
      subtitle: localizedOptional(section.subtitle, table),
      iconName: section.iconName,
      iconEmoji: section.iconEmoji,
      dataSource: section.dataSource,
      controls: section.controls
          .map((control) => _localizeControl(control, table))
          .toList(),
      actions: section.actions
          .map((action) => _localizeAction(action, table))
          .toList(),
    );

ControlSpec _localizeControl(ControlSpec control, Map<String, String> table) =>
    ControlSpec(
      id: control.id,
      label: localized(control.label, table),
      kind: control.kind,
      value: control.value,
      placeholder: localizedOptional(control.placeholder, table),
      tooltip: localizedOptional(control.tooltip, table),
      options: control.options
          .map((option) => ControlOption(
                id: option.id,
                title: localized(option.title, table),
                selected: option.selected,
                status: option.status,
                group: localizedOptional(option.group, table),
              ))
          .toList(),
      columns: control.columns
          .map((column) => ListColumnSpec(
              id: column.id, title: localized(column.title, table)))
          .toList(),
      rows: control.rows.map((row) => _localizeRow(row, table)).toList(),
      items: control.items
          .map((item) => ListItemSpec(item.values
              .map((key, value) => MapEntry(key, localized(value, table)))))
          .toList(),
      rowTemplate: control.rowTemplate == null
          ? null
          : _localizeRow(control.rowTemplate!, table),
      rowActions: control.rowActions
          .map((action) => _localizeAction(action, table))
          .toList(),
      settings: control.settings
          .map((setting) => _localizeSetting(setting, table))
          .toList(),
      configFile: control.configFile,
      dataSource: control.dataSource,
    );

ListRowSpec _localizeRow(ListRowSpec row, Map<String, String> table) =>
    ListRowSpec(
      id: row.id,
      title: localizedOptional(row.title, table),
      values: row.values,
      status: localizedOptional(row.status, table),
      tooltip: localizedOptional(row.tooltip, table),
      tags: row.tags
          .map((tag) => TagSpec(
              id: tag.id, title: localized(tag.title, table), style: tag.style))
          .toList(),
    );

ActionSpec _localizeAction(ActionSpec action, Map<String, String> table) =>
    ActionSpec(
      id: action.id,
      title: localized(action.title, table),
      tooltip: localizedOptional(action.tooltip, table),
      iconName: action.iconName,
      iconEmoji: action.iconEmoji,
      iconOnly: action.iconOnly,
      role: action.role,
      destructive: action.destructive,
      visibleWhen: action.visibleWhen,
      disabledWhen: action.disabledWhen,
      disabledTooltip: localizedOptional(action.disabledTooltip, table),
      precheck: action.precheck == null
          ? null
          : ActionPrecheckSpec(
              diskSpaceGB: action.precheck!.diskSpaceGB,
              diskSpacePath: action.precheck!.diskSpacePath,
              warningMessage:
                  localizedOptional(action.precheck!.warningMessage, table),
            ),
      confirm: action.confirm == null
          ? null
          : ActionConfirmationSpec(
              title: localized(action.confirm!.title, table),
              message: localizedOptional(action.confirm!.message, table),
              confirmButtonTitle:
                  localized(action.confirm!.confirmButtonTitle, table),
              cancelButtonTitle:
                  localized(action.confirm!.cancelButtonTitle, table),
              requiredText: action.confirm!.requiredText,
              prompt: localizedOptional(action.confirm!.prompt, table),
            ),
      command: action.command,
    );

ConfigSettingSpec _localizeSetting(
        ConfigSettingSpec setting, Map<String, String> table) =>
    ConfigSettingSpec(
      id: setting.id,
      kind: setting.kind,
      key: setting.key,
      label: localized(setting.label, table),
      value: setting.value,
      placeholder: localizedOptional(setting.placeholder, table),
      tooltip: localizedOptional(setting.tooltip, table),
      options: setting.options
          .map((option) => ControlOption(
                id: option.id,
                title: localized(option.title, table),
                selected: option.selected,
                status: option.status,
                group: localizedOptional(option.group, table),
              ))
          .toList(),
      dataSource: setting.dataSource,
    );

int _findUnescapedEquals(String line) {
  var quoted = false;
  var escaped = false;
  for (var index = 0; index < line.length; index += 1) {
    final character = line[index];
    if (escaped) {
      escaped = false;
    } else if (character == '\\') {
      escaped = true;
    } else if (character == '"') {
      quoted = !quoted;
    } else if (character == '=' && !quoted) {
      return index;
    }
  }
  return -1;
}

int _findClosingQuote(String value) {
  var escaped = false;
  for (var index = 1; index < value.length; index += 1) {
    final character = value[index];
    if (escaped) {
      escaped = false;
    } else if (character == '\\') {
      escaped = true;
    } else if (character == '"') {
      return index;
    }
  }
  return -1;
}

String _unquoteKey(String key) => key.startsWith('"') && key.endsWith('"')
    ? _unescapeTomlString(key.substring(1, key.length - 1))
    : key;

String _unescapeTomlString(String value) => value
    .replaceAll(r'\"', '"')
    .replaceAll(r'\n', '\n')
    .replaceAll(r'\r', '\r')
    .replaceAll(r'\t', '\t')
    .replaceAll(r'\\', '\\');

File: exp-platform/kotlin/compose/shared/src/main/kotlin/dev/guiforcli/compose/runtime/Localization.kt
package dev.guiforcli.compose.runtime

import dev.guiforcli.compose.model.ActionConfirmationSpec
import dev.guiforcli.compose.model.ActionPrecheckSpec
import dev.guiforcli.compose.model.ActionSpec
import dev.guiforcli.compose.model.BundleManifest
import dev.guiforcli.compose.model.BundlePage
import dev.guiforcli.compose.model.ConfigSettingSpec
import dev.guiforcli.compose.model.ControlOption
import dev.guiforcli.compose.model.ControlSpec
import dev.guiforcli.compose.model.ExitCodeReferenceEntry
import dev.guiforcli.compose.model.ListColumnSpec
import dev.guiforcli.compose.model.ListRowSpec
import dev.guiforcli.compose.model.PageSection
import dev.guiforcli.compose.model.SetupSpec
import dev.guiforcli.compose.model.SetupStep
import dev.guiforcli.compose.model.TagSpec

fun parseTomlStringTable(text: String): Map<String, String> {
    val values = linkedMapOf<String, String>()
    for (rawLine in text.lineSequence()) {
        val line = rawLine.trim()
        if (line.isEmpty() || line.startsWith("#")) {
            continue
        }
        val split = line.indexOf('=')
        if (split <= 0) {
            continue
        }
        val key = unquoteTomlString(line.substring(0, split).trim())
        val value = unquoteTomlString(line.substring(split + 1).trim())
        values[key] = value
    }
    return values
}

fun BundleManifest.localized(strings: Map<String, String>): BundleManifest =
    copy(
        displayName = strings.localize(displayName) ?: displayName,
        summary = strings.localize(summary) ?: summary,
        setup = setup.localized(strings),
        exitCodeReference = exitCodeReference.map { it.localized(strings) },
        pages = pages.map { it.localized(strings) },
    )

fun Map<String, String>.localize(value: String?): String? = value?.let { this[it] ?: it }

private fun SetupSpec.localized(strings: Map<String, String>): SetupSpec =
    copy(steps = steps.map { it.localized(strings) })

private fun SetupStep.localized(strings: Map<String, String>): SetupStep =
    copy(label = strings.localize(label) ?: label)

private fun ExitCodeReferenceEntry.localized(strings: Map<String, String>): ExitCodeReferenceEntry =
    copy(title = strings.localize(title) ?: title, summary = strings.localize(summary) ?: summary)

private fun BundlePage.localized(strings: Map<String, String>): BundlePage =
    copy(
        title = strings.localize(title) ?: title,
        summary = strings.localize(summary) ?: summary,
        sidebarGroup = strings.localize(sidebarGroup),
        sections = sections.map { it.localized(strings) },
    )

private fun PageSection.localized(strings: Map<String, String>): PageSection =
    copy(
        title = strings.localize(title),
        subtitle = strings.localize(subtitle),
        controls = controls.map { it.localized(strings) },
        actions = actions.map { it.localized(strings) },
    )

private fun ControlSpec.localized(strings: Map<String, String>): ControlSpec =
    copy(
        label = strings.localize(label) ?: label,
        placeholder = strings.localize(placeholder),
        tooltip = strings.localize(tooltip),
        options = options.map { it.localized(strings) },
        columns = columns.map { it.localized(strings) },
        rows = rows.map { it.localized(strings) },
        rowTemplate = rowTemplate?.localized(strings),
        rowActions = rowActions.map { it.localized(strings) },
        settings = settings.map { it.localized(strings) },
    )

private fun ControlOption.localized(strings: Map<String, String>): ControlOption =
    copy(title = strings.localize(title) ?: title, subtitle = strings.localize(subtitle))

private fun ListColumnSpec.localized(strings: Map<String, String>): ListColumnSpec =
    copy(title = strings.localize(title) ?: title)

private fun ListRowSpec.localized(strings: Map<String, String>): ListRowSpec =
    copy(
        title = strings.localize(title),
        subtitle = strings.localize(subtitle),
        status = strings.localize(status),
        tags = tags.map { it.localized(strings) },
    )

private fun TagSpec.localized(strings: Map<String, String>): TagSpec =
    copy(title = strings.localize(title) ?: title)

private fun ActionSpec.localized(strings: Map<String, String>): ActionSpec =
    copy(
        title = strings.localize(title) ?: title,
        tooltip = strings.localize(tooltip),
        disabledTooltip = strings.localize(disabledTooltip),
        precheck = precheck?.localized(strings),
        confirm = confirm?.localized(strings),
    )

private fun ActionPrecheckSpec.localized(strings: Map<String, String>): ActionPrecheckSpec =
    copy(warningMessage = strings.localize(warningMessage))

private fun ActionConfirmationSpec.localized(strings: Map<String, String>): ActionConfirmationSpec =
    copy(
        title = strings.localize(title) ?: title,
        message = strings.localize(message),
        confirmButtonTitle = strings.localize(confirmButtonTitle) ?: confirmButtonTitle,
        cancelButtonTitle = strings.localize(cancelButtonTitle) ?: cancelButtonTitle,
        requiredText = strings.localize(requiredText) ?: requiredText,
        prompt = strings.localize(prompt),
    )

private fun ConfigSettingSpec.localized(strings: Map<String, String>): ConfigSettingSpec =
    copy(
        label = strings.localize(label) ?: label,
        placeholder = strings.localize(placeholder),
        tooltip = strings.localize(tooltip),
        options = options.map { it.localized(strings) },
    )

private fun unquoteTomlString(value: String): String {
    val trimmed = value.trim().removeSuffix(",")
    if (trimmed.length < 2 || trimmed.first() != '"' || trimmed.last() != '"') {
        return trimmed
    }
    val body = trimmed.substring(1, trimmed.length - 1)
    return buildString {
        var index = 0
        while (index < body.length) {
            val ch = body[index]
            if (ch == '\\' && index + 1 < body.length) {
                append(
                    when (val escaped = body[index + 1]) {
                        'n' -> '\n'
                        'r' -> '\r'
                        't' -> '\t'
                        '"' -> '"'
                        '\\' -> '\\'
                        else -> escaped
                    },
                )
                index += 2
            } else {
                append(ch)
                index += 1
            }
        }
    }
}
